### PR TITLE
can not prefix with ``this.``

### PR DESCRIPTION
**Pull request by gqqnbig**
I'm a newbie, but I tried the following code

```
pragma solidity 0.8.4;

contract Add {

    uint x;

    constructor(uint a) public {
        this.x = a;
    }

}
```

Compiler gives error:

```
CompileError: TypeError: Member "x" not found or not visible after argument-dependent lookup in contract Add.
  --> /B/add/contracts/add.sol:14:9:
   |
14 |         this.x = a;
   |         ^^^^^^
```

If I remove `this.`, the code compiles successfully.

I checked the documentation on https://docs.soliditylang.org/en/develop/introduction-to-smart-contracts.html, which says "do not need the prefix this". This statement is confusing because it implies people are still able to prefix while the reality is prefixing fails the compiler.

As a result, I made the PR to make the statement clearer.

### CODE REVIEW

**Comment by Marenz** [comment]
The matter is not as clear as it might look. You can actually use `this` but only for `public` variables (for `external` you _need_ this). For `internal` (the default if nothing specified, the case here) you can't use `this` again. 

`this` basically imitates an external access to the variable. 

A more correct way to fix the documentation could be this:

> To access an ``internal`` state variable, you do not need the prefix ``this.`` as is common in
other languages.

**Comment by chriseth** [comment]
I think "do not need" is still clearer than "cannot use", because the latter does not state how you can access a state variable.

File: docs/introduction-to-smart-contracts.rst [hunk 1]
```diff
@@ -51,7 +51,7 @@ code that manages the database. In this example, the contract defines the
 functions ``set`` and ``get`` that can be used to modify
 or retrieve the value of the variable.
 
-To access a state variable, you do not need the prefix ``this.`` as is common in
+To access a state variable, you can not prefix with ``this.`` as is common in
 other languages.
```
Comment (cameel) [inline docs/introduction-to-smart-contracts.rst]: How about this?
```markdown
State variables can be accessed without the ``this.`` prefix.
Note that, unlike in some other languages, omitting it is not just a matter of style.
Accessing a variable through a contract instance results in an external call,
which is more expensive and also cannot be used to modify it.
```

Comment (gqqnbig) [reply docs/introduction-to-smart-contracts.rst]: I like the first half as it says the usage of `this` is unlike other languages and omitting it is not just a matter of style. The second half is too esoteric.

How about this?
```
State variables can be accessed without the ``this.`` prefix.
Note that, unlike in some other languages, omitting it is not just a matter of style.
You can actually use `this` but only for public variables (for external you need `this`). 
For internal, you can't use `this`.
```

Comment (cameel) [reply docs/introduction-to-smart-contracts.rst]: Well, if you think that going into details in unnecessary then we could just change it to this:
```rst
To access a state variable, you do not typically add the ``this.`` prefix.
Unlike in some other languages, omitting it is not just a matter of style,
it results in a completely different operation.
```

But if you want it to say why, here's another version:
```rst
To access a state variable, you do not typically add the ``this.`` prefix.
Unlike in some other languages, omitting it is not just a matter of style.
Using ``this`` you can only access its :ref:`getter function <getter-functions>`
that is only provided for public state variables.
```

Note that `` :ref:`getter function <getter-functions>` `` is RST syntax for a link to [Contracts > Getter Functions](https://docs.soliditylang.org/en/latest/contracts.html#getter-functions).

**Comment by chriseth** [comment]
I think we should also clarify that this will result in an external call: "State variables and all other members of the current contract are accessed directly, without the `this.` prefix. For functions and public state variables, you can use `this.`, but you will obtain the externally-facing part of the item (external function or getter function) and calling it will result in a costly external call."

File: docs/introduction-to-smart-contracts.rst (hunk 1 above)
Comment (gqqnbig) [reply docs/introduction-to-smart-contracts.rst]: ```... it results in a completely different operation. ``` is good for me!

**Comment by cameel** [comment]
@chriseth I think that keeping it simple might be better after all. This is at the very beginning of "Introduction to Smart Contracts" which, I guess, is why @gqqnbig considers my version esoteric :) This is also why I scaled it back to something expressed in simpler terms.

File: docs/introduction-to-smart-contracts.rst (hunk 1 above)
Comment (cameel) [reply docs/introduction-to-smart-contracts.rst]: So could you change the PR to contain version of your choice?

**Comment by gqqnbig** [comment]
Hi all, I adopted cameel's version, please check.

File: docs/introduction-to-smart-contracts.rst [hunk 2]
```diff
@@ -51,8 +51,9 @@ code that manages the database. In this example, the contract defines the
 functions ``set`` and ``get`` that can be used to modify
 or retrieve the value of the variable.
 
-To access a state variable, you can not prefix with ``this.`` as is common in
-other languages.
+To access a state variable, you do not typically add the ``this.`` prefix.
+Unlike in some other languages, omitting it is not just a matter of style,
+it results in a completely different operation.
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]:
```suggestion
To access a member (like a state variable) of the current contract, you do not typically add the ``this.`` prefix,
you just access it directly via its name.
Unlike in some other languages, omitting it is not just a matter of style,
it results in a completely different way to access the member, but more on this later.
```

Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst]: Not sure if this is better...

Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst]: Sorry that this discussion is dragging on for so long, but the early chapters in the documentation are often the most difficult ones.

Comment (cameel) [reply docs/introduction-to-smart-contracts.rst]: I think it's fine too but let's see what is @gqqnbig's opinion.

**Comment by gqqnbig** [comment]
I accepted the suggestion. Not sure why the test failed.

**Review by cameel — APPROVED** [review]
The test failure is unrelated. It has already been fixed in #11529. You can rebase the PR on develop if you want to see all tests pass but I think it's pretty clear that it's not breaking anything :)

**Comment by gqqnbig** [comment]
Thank you everyone!
